### PR TITLE
bpo-47049: Fix incorrect shutil.copytree() behaviour with symlinks

### PR DESCRIPTION
**Pull request by vajdaz**
shutil.copytree incorrectly does not copy symlink contents if called
with symlink=False and ignore_dangling_symlinks=True.

The wrong behaviour can be reproduced like this:
```
$ tree
.
└── a
    ├── a.txt
    └── b
        └── a.txt -> ../a.txt
```
```$ python3 -c "import shutil;shutil.copytree('a/b', 'c', symlinks=False, ignore_dangling_symlinks=True)"```

As a result directoy c will be created but it will remain empty.
Expected result is a file c/a.txt with the contents of a/b/a.txt.

https://bugs.python.org/issue47049

### CODE REVIEW

File: Lib/shutil.py [hunk 1]
```diff
@@ -483,6 +483,12 @@ def _copytree(entries, src, dst, symlinks, ignore, copy_function,
                     os.symlink(linkto, dstname)
                     copystat(srcobj, dstname, follow_symlinks=not symlinks)
                 else:
+                    # If the symlink is relative, linkto has to be normalized,
+                    # otherwise os.path.exists() will incorrectly report that
+                    # the link is dangling when the current directoy is not
```
Comment (arhadthedev) [inline Lib/shutil.py]:
```suggestion
                    # the link is dangling when the current directory is not
```

**Comment by vajdaz** [comment]
Something in the CLA signing check process changed (old checking service unavailable), and therefore the check fails although I signed the CLA. How can this be fixed? Do I have to do something, or does a a maintainer have to do something?

**Comment by MaxwellDupre** [comment]
I have tested with test_shutil (and the example given) and it looks fine, but it looks like there are tests (in test_shutil) which are negative due to this bug.
Could you check to see if some tests need to be changed? Possibly test_copytree_dangling_symlinks.
If none then can you add a test for this change?

**Comment by vajdaz** [comment]
I have added a unit test to the fix. That revealed another issue. When dangling symlinks do exist in this situation, an error was thrown, because the ignore_dangling_symlinks argument of the copytree function was not propagated to recursive calls. This is fixed by the last commit, too.

**Comment by vajdaz** [comment]
@MaxwellDupre can you please have a look on this again?

**Review by MaxwellDupre — APPROVED** [review]
LGTM: module and given tests run ok.
However, there is a lot of comments Lib/test/test_shutil.py esp lines 771-778. Most comment for tests are only a few lines.
You already hae ref'ed the Issue 47049 so people can go look that up if they want or check the PR. 
Hence, you description lines 779 onwards could be a little shorter too.

**Comment by vajdaz** [comment]
> LGTM: module and given tests run ok. However, there is a lot of comments Lib/test/test_shutil.py esp lines 771-778. Most comment for tests are only a few lines. You already hae ref'ed the Issue 47049 so people can go look that up if they want or check the PR. Hence, you description lines 779 onwards could be a little shorter too.
@MaxwellDupre I cut the comment.

**Comment by vajdaz** [comment]
ping @MaxwellDupre

**Review by MaxwellDupre — APPROVED** [review]
Better but still a few extra words.

**Comment by vajdaz** [comment]
> Better but still a few extra words.

@MaxwellDupre :) How about now?

**Comment by vajdaz** [comment]
Seems that this has been solved in #22937 and it is also a duplicate of #82704. Therefore I close this PR.

**Comment by dabuahoid** [comment]
The issue is still present in Python 3.12.6 and also an open case exists: https://github.com/python/cpython/issues/91205#issue-1199077196

**Pull request closed without merging**